### PR TITLE
feat(client): add hard compaction option preserving system segment

### PR DESCRIPTION
**Pull request by HUQIANTAO**
## Summary

Add a hard compaction mode that replaces the middle conversation history with a single summary message while preserving the system prompt segment and the last N recent messages (default 8). This is opt-in via the `hard_enabled` configuration flag.

## Background

The existing compaction system uses a soft-seam approach: it appends `<archived_context>` summary blocks to the message array without removing old messages. This preserves all verbatim history but keeps growing the message array. When the context window is under pressure, the soft-seam approach may not free enough budget.

Hard compaction is more aggressive: it replaces the middle history entirely with a summary message. The trade-off is that old messages are lost (they're summarized), but the system prompt segment remains byte-identical, so the prefix cache stays hot after compaction.

### Message structure after hard compaction

```text
[summary_message, recent_message_1, recent_message_2, ..., recent_message_8]
```

The system prompt (stored separately in `session.system_prompt`) is never touched. The summary message replaces the middle history, and the recent tail is preserved verbatim.

### Configuration

Two new fields in `CompactionConfig`:

- `hard_enabled: bool` (default: `false`) — opt-in flag
- `hard_keep_recent: usize` (default: `8`) — number of recent messages to preserve

### API

- `plan_hard_compaction(messages, keep_recent) -> Option<HardCompactionPlan>`: Plan which messages to summarize and which to keep.
- `compact_hard_safe(client, messages, config, workspace) -> Result<HardCompactionResult>`: Execute the hard compaction using the existing LLM summarization infrastructure.

### Summary format

The summary is wrapped in `<hard_compaction_summary>` tags so the model can distinguish it from verbatim history:

```text
<hard_compaction_summary>
[LLM-generated summary of the middle history]
</hard_compaction_summary>
```

## Why this matters

Hard compaction is the highest-leverage tool for freeing context budget when the window is under pressure. Unlike soft seams (which keep growing the message array), hard compaction replaces the middle history entirely. The system prompt segment (which is the largest stable block) is never touched, so the prefix cache remains hot after compaction.

This is especially valuable for long sessions where the context window is exhausted: instead of starting a new cycle (which resets the prefix cache), hard compaction can free enough budget to continue the current cycle with the prefix cache intact.

## Testing

3 new unit tests:

- `plan_hard_compaction_returns_none_when_too_few_messages`: Verify graceful handling of insufficient messages.
- `plan_hard_compaction_preserves_recent_tail`: Verify the recent tail is preserved correctly.
- `hard_compaction_config_defaults`: Verify the default configuration values.

All 59 compaction tests pass.

## Files changed

- `crates/tui/src/compaction.rs`: Add hard compaction types, functions, and tests (223 lines added)

### CODE REVIEW

File: crates/tui/src/compaction.rs [hunk 1]
```diff
@@ -1027,6 +1107,100 @@ pub async fn compact_messages_safe(
...
+///
+/// # Returns
+/// `Ok(HardCompactionResult)` with the new message list, or `Err` if
+/// summarization fails after retries.
+pub async fn compact_hard_safe(
+    client: &DeepSeekClient,
+    messages: &[Message],
+    config: &CompactionConfig,
+    workspace: Option<&Path>,
+) -> Result<HardCompactionResult> {
```
[the remark this answers is not in the document]
Comment (carloslindhstrom-code) [reply crates/tui/src/compaction.rs:1117]: Kj

**Comment by Hmbown** [comment]
Thanks @HUQIANTAO. I reviewed this for v0.9.0. I am going to defer it for this release train rather than merge it directly.

The hard-compaction direction is useful, but the current branch adds a separate hard-compactor path that is not wired into the existing compaction flow, can split tool-call/result pairs by cutting on message index, and bypasses the current compaction contracts around tool-pair enforcement, pinned working sets, anchors, and cache-inspection telemetry. It also adds an extra summary system prompt, which works against the prefix-stability goal.

A future harvest should reuse the existing `plan_compaction` / tool-pair machinery in `crates/tui/src/compaction.rs`, avoid extra system-prefix churn, add a real setting before enabling it, and include tests for tool-call/result atomicity, tail boundaries, anchors/working-set preservation, and cache/turn metadata stability. We should credit #2522 if that lands.

**Comment by Hmbown** [comment]
Thanks again @HUQIANTAO. I harvested the safe planning layer from this proposal into #2838, now merged on the v0.9 stewardship branch as `d784f1e`.

What landed:
- a default-disabled `HardCompactionConfig` marker;
- a pure `plan_hard_compaction` helper;
- planner tests for too-short histories, recent-tail preservation, and the tool-call/result boundary risk raised in review.

The harvest intentionally reuses the existing `plan_compaction` machinery so working-set pins, recent-tail preservation, user-text fallback, and tool-call/result pair guarantees stay intact. It does not enable hard compaction, add user config, or add a message-rewrite/LLM execution path yet.

Verification on the maintainer PR: `cargo test -p codewhale-tui compaction --locked`, `cargo clippy -p codewhale-tui --bin codewhale-tui --locked -- -D warnings`, `cargo check --workspace --all-features --locked`, `cargo fmt --all -- --check`, `./scripts/release/check-versions.sh`, `./scripts/release/check-ohos-deps.sh`, and `git diff --check`.

I am leaving this PR open for now because the broader hard-compaction execution path still needs the runtime semantics from the earlier review: no tool-pair splitting, no unexpected system-prefix churn, and stronger tests around anchors/working-set preservation and cache metadata stability.

**Comment by Hmbown** [comment]
Closing with harvest credit, @HUQIANTAO — the hard-compaction option preserving the system segment was harvested into #2838 and is on main. Thank you for the original design and implementation; this thread remains the reference for the approach.
